File: week_2/src/server_pkg/src/server_sub_node.cpp
#include <ros/ros.h>
#include "server_pkg/addint.h"
/*
	客户端：提交两个整数，并处理响应
		1.包含头文件
		2.setlocale，初始化
		3.创建大管家
		4.创建客户端对象
		5.提交请求并处理响应
*/
int main(int argc, char *argv[])
{
    // argc:代表传入参数的总数量
    // 本案例共三个参数：程序名称，num1,num2
    if(argc != 3)
    {
        ROS_WARN("传入参数个数不对！");
        return 0;
    }

    setlocale(LC_ALL,"");
    ros::init(argc,argv,"server_sub_node");
    ros::NodeHandle nh;

    // 4.创建客户端对象
    ros::ServiceClient client = nh.serviceClient<server_pkg::addint>("/addint");

    // 5.提交请求并处理响应
    // 5-1.组织请求

    // 获取服务对象
    server_pkg::addint ai;
    // 获取服务请求对象
    ai.request.num1 = atoi(argv[1]);
    ai.request.num2 = atoi(argv[2]);

    // 5-2.处理响应

    // 等待服务端启动
    client.waitForExistence();

    // 客户端对象通过call函数查看是否成功处理请求
    bool flag = client.call(ai);

    if(flag)
    {
        ROS_WARN("客户端接受响应");
        ROS_INFO("响应sums = %d",ai.response.sums);
    }
    else
    {
        ROS_WARN("客户端响应失败");
    }
    return 0;
}

File: week_2/src/server_pkg/scripts/server_sub_node.py
#!/usr/bin/env python3
#encoding=utf-8

import rospy
from server_pkg.srv import*
import sys

if __name__ == "__main__":
    # 优化实现，参数的动态提交
    # 需要包含库文件sys

    # 判断传入的参数个数
    if len(sys.argv) != 3:
        rospy.loginfo("传入参数个数不对！")
        sys.exit(1)


    # 初始化
    rospy.init_node("server_sub_node")
    
    # 获取客户端对象
    client = rospy.ServiceProxy("/addint",addint)

    # 组织请求，获取响应
    num1 = int(sys.argv[1])
    num2 = int(sys.argv[2])

    # 优化实现
    # 2.客户端挂起
    # 2-1通过客户端对象挂起
    client.wait_for_service()

    # 2-2 通过rospy挂起
    # 传入的参数是要等待的服务话题
    rospy.wait_for_service("/addint")
    
    response = client.call(num1,num2)

    # 处理请求
    rospy.loginfo("响应的数据sums=%d",response.sums)
    